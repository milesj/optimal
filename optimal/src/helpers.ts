import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
