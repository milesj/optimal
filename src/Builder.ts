import isObject from './isObject';
import {
  SupportedType,
  CheckerCallback,
  CustomCallback,
  OptimalOptions,
  FuncOf,
  DefaultValue,
  DefaultValueFactory,
} from './types';

export interface TemporalStruct {
  [key: string]: unknown;
}

export default class Builder<T> {
  defaultValue?: T;

  type: SupportedType;

  protected checks: CheckerCallback[] = [];

  protected currentStruct: object = {};

  protected defaultValueFactory?: DefaultValueFactory<T>;

  protected deprecatedMessage: string = '';

  protected errorMessage: string = '';

  protected isNever: boolean = false;

  protected isNullable: boolean = false;

  protected isRequired: boolean = false;

  protected noErrorPrefix: boolean = false;

  protected options: OptimalOptions = {};

  constructor(type: SupportedType, defaultValue: DefaultValue<T>, bypassFactory: boolean = false) {
    if (__DEV__) {
      this.invariant(
        typeof defaultValue !== 'undefined',
        `A default value for type "${type}" is required.`,
      );

      this.addCheck(this.checkType);
    }

    if (typeof defaultValue === 'function' && !bypassFactory) {
      this.defaultValueFactory = defaultValue as DefaultValueFactory<T>;
    } else {
      this.defaultValue = defaultValue as T;
    }

    this.type = type;
  }

  /**
   * Map a list of names that must be defined alongside this field.
   */
  and(...keys: string[]): this {
    if (__DEV__) {
      this.invariant(keys.length > 0, 'AND requires a list of field names.');

      this.addCheck(path => {
        const checkedKeys = [this.key(path), ...keys];
        const struct = this.currentStruct as TemporalStruct;
        const undefs = checkedKeys.filter(
          key => typeof struct[key] === 'undefined' || struct[key] === null,
        );

        // Only error once one of the struct is defined
        if (undefs.length === checkedKeys.length) {
          return;
        }

        this.invariant(
          undefs.length === 0,
          `All of these fields must be defined: ${checkedKeys.join(', ')}`,
        );
      });
    }

    return this;
  }

  /**
   * Set a callback to run custom logic.
   */
  custom<S = object>(callback: CustomCallback<T, S>): this {
    if (__DEV__) {
      this.invariant(
        typeof callback === 'function',
        'Custom blueprints require a validation function.',
      );

      this.addCheck((path, value) => {
        try {
          callback(value, (this.currentStruct as unknown) as S);
        } catch (error) {
          this.invariant(false, error.message, path);
        }
      });
    }

    return this;
  }

  /**
   * Set a message to log when this field is present.
   */
  deprecate(message: string): this {
    if (__DEV__) {
      this.invariant(
        typeof message === 'string' && !!message,
        'A non-empty string is required for deprecated messages.',
      );
    }

    this.deprecatedMessage = message;

    return this;
  }

  /**
   * Throw an error if the condition is falsy.
   */
  invariant(condition: boolean, message: string, path: string = '') {
    if (__DEV__) {
      if (condition) {
        return;
      }

      const { file, name } = this.options;
      const error = this.errorMessage || message;
      let prefix = '';

      if (path) {
        if (name) {
          prefix += `Invalid ${name} field "${path}"`;
        } else {
          prefix += `Invalid field "${path}"`;
        }
      } else if (name) {
        prefix += name;
      }

      if (file) {
        prefix += ` in ${file}`;
      }

      if (prefix && !this.noErrorPrefix) {
        throw new Error(`${prefix}. ${error}`);
      } else {
        throw new Error(error);
      }
    }
  }

  /**
   * Set a custom error message for all checks.
   */
  message(message: string): this {
    if (__DEV__) {
      this.invariant(
        typeof message === 'string' && !!message,
        'A non-empty string is required for custom messages.',
      );
    }

    this.errorMessage = message;

    return this;
  }

  /**
   * Field should never be used.
   */
  never(): Builder<never> {
    this.defaultValue = (undefined as unknown) as T;
    this.isNever = true;

    return (this as unknown) as Builder<never>;
  }

  /**
   * Disallow null values.
   */
  notNullable(): Builder<NonNullable<T>> {
    this.isNullable = false;

    return (this as unknown) as Builder<NonNullable<T>>;
  }

  /**
   * Allow null values.
   */
  nullable(): Builder<T | null> {
    this.isNullable = true;

    return (this as unknown) as Builder<T | null>;
  }

  /**
   * Mark a field as only the default value can be used.
   */
  only(): this {
    if (__DEV__) {
      this.invariant(
        // eslint-disable-next-line valid-typeof
        typeof this.defaultValue === this.type,
        `Only requires a default ${this.type} value.`,
      );

      this.addCheck((path, value) => {
        this.invariant(
          value === this.defaultValue,
          `Value may only be "${String(this.defaultValue)}".`,
          path,
        );
      });
    }

    return this;
  }

  /**
   * Map a list of field names that must have at least 1 defined.
   */
  or(...keys: string[]): this {
    if (__DEV__) {
      this.invariant(keys.length > 0, 'OR requires a list of field names.');

      this.addCheck(path => {
        const orKeys = [this.key(path), ...keys];
        const struct = this.currentStruct as TemporalStruct;
        const defs = orKeys.filter(
          key => typeof struct[key] !== 'undefined' && struct[key] !== null,
        );

        this.invariant(
          defs.length > 0,
          `At least one of these fields must be defined: ${orKeys.join(', ')}`,
        );
      });
    }

    return this;
  }

  /**
   * Require an object property to be explicitly defined.
   */
  required(state: boolean = true): this {
    this.isRequired = state;

    return this;
  }

  /**
   * Run all validation checks that have been enqueued.
   */
  // eslint-disable-next-line complexity
  runChecks(
    path: string,
    initialValue: T | undefined,
    struct: object,
    options: OptimalOptions = {},
  ): T | null {
    this.currentStruct = struct;
    this.options = options;

    if (this.defaultValueFactory) {
      this.defaultValue = this.defaultValueFactory(struct);
    }

    let value = initialValue;

    // Handle undefined
    if (typeof value === 'undefined') {
      if (!this.isRequired) {
        value = this.defaultValue;
      } else if (__DEV__) {
        this.invariant(false, 'Field is required and must be defined.', path);
      }
    } else if (__DEV__) {
      if (this.deprecatedMessage) {
        // eslint-disable-next-line no-console
        console.info(`Field "${path}" is deprecated. ${this.deprecatedMessage}`);
      }

      if (this.isNever) {
        this.invariant(false, 'Field should never be used.', path);
      }
    }

    // Handle null
    if (value === null) {
      if (this.isNullable) {
        return value;
      }

      if (__DEV__) {
        this.invariant(false, 'Null is not allowed.', path);
      }
    }

    // Run all checks against the value
    if (__DEV__) {
      this.checks.forEach(checker => {
        const result = checker.call(this, path, value);

        if (typeof result !== 'undefined') {
          value = result as T;
        }
      });
    }

    return value!;
  }

  /**
   * Return a human readable type name.
   */
  typeAlias(): string {
    return this.type;
  }

  /**
   * Map a list of field names that must not be defined alongside this field.
   */
  xor(...keys: string[]): this {
    if (__DEV__) {
      this.invariant(keys.length > 0, 'XOR requires a list of field names.');

      this.addCheck(path => {
        const xorKeys = [this.key(path), ...keys];
        const struct = this.currentStruct as TemporalStruct;
        const defs = xorKeys.filter(
          key => typeof struct[key] !== 'undefined' && struct[key] !== null,
        );

        this.invariant(
          defs.length === 1,
          `Only one of these fields may be defined: ${xorKeys.join(', ')}`,
        );
      });
    }

    return this;
  }

  /**
   * Add a checking function with optional arguments.
   */
  protected addCheck(checker: CheckerCallback): this {
    if (__DEV__) {
      this.checks.push(checker);
    }

    return this;
  }

  /**
   * Validate the type of value.
   */
  protected checkType(path: string, value: T) {
    if (__DEV__) {
      switch (this.type) {
        case 'array':
          this.invariant(Array.isArray(value), 'Must be an array.', path);
          break;

        case 'custom':
        case 'instance':
        case 'union':
          // Handle in the sub-class
          break;

        case 'object':
        case 'shape':
          this.invariant(isObject(value), 'Must be a plain object.', path);
          break;

        default:
          // eslint-disable-next-line valid-typeof
          this.invariant(typeof value === this.type, `Must be a ${this.type}.`, path);
          break;
      }
    }
  }

  /**
   * Return true if the value matches the default value and the builder is optional.
   */
  protected isOptionalDefault(value: unknown): boolean {
    return !this.isRequired && value === this.defaultValue;
  }

  /**
   * Return the current key from a path.
   */
  protected key(path: string): string {
    const index = path.lastIndexOf('.');

    return index > 0 ? path.slice(index + 1) : path;
  }
}

export function custom<T, S = object>(
  callback: CustomCallback<T, S>,
  defaultValue: DefaultValue<T>,
) /* infer */ {
  return new Builder<T>('custom', defaultValue).custom(callback);
}

export function func<T = FuncOf>(defaultValue: T | null = null) /* infer */ {
  return new Builder<T | null>('function', defaultValue, true).nullable();
}
