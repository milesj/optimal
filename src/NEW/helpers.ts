import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
