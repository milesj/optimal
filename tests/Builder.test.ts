import Builder, { custom, func } from '../src/Builder';
import { runChecks } from './helpers';

describe('Builder', () => {
  let builder: Builder<unknown>;

  beforeEach(() => {
    builder = new Builder('string', 'foo');
  });

  it('errors if default value is undefined', () => {
    expect(() => {
      builder = new Builder('string', undefined);
    }).toThrowErrorMatchingSnapshot();
  });

  it('sets the type and default value', () => {
    expect(builder.type).toBe('string');
    expect(builder.defaultValue).toBe('foo');
  });

  describe('and()', () => {
    beforeEach(() => {
      builder.and('bar', 'baz');
    });

    it('errors if no keys are defined', () => {
      expect(() => {
        builder.and();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if not all options are defined', () => {
      expect(() => {
        runChecks(builder, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
            baz: 'c',
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if all are defined', () => {
      expect(() => {
        runChecks(builder, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
            bar: 'b',
            baz: 'c',
          },
        });
      }).not.toThrow();
    });
  });

  describe('checkType()', () => {
    describe('array', () => {
      it('allows arrays', () => {
        builder.type = 'array';

        expect(() => {
          runChecks(builder, []);
        }).not.toThrow();
      });

      it('errors on non-arrays', () => {
        builder.type = 'array';

        expect(() => {
          runChecks(builder, 123);
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('boolean', () => {
      it('allows booleans', () => {
        builder.type = 'boolean';

        expect(() => {
          runChecks(builder, true);
        }).not.toThrow();
      });

      it('errors on non-booleans', () => {
        builder.type = 'boolean';

        expect(() => {
          runChecks(builder, 123);
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('function', () => {
      it('allows functions', () => {
        builder.type = 'function';

        expect(() => {
          runChecks(builder, () => {});
        }).not.toThrow();
      });

      it('errors on non-functions', () => {
        builder.type = 'function';

        expect(() => {
          runChecks(builder, 'foo');
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('number', () => {
      it('allows numbers', () => {
        builder.type = 'number';

        expect(() => {
          runChecks(builder, 123);
        }).not.toThrow();
      });

      it('errors on non-numbers', () => {
        builder.type = 'number';

        expect(() => {
          runChecks(builder, 'foo');
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('object', () => {
      it('allows objects', () => {
        builder.type = 'object';

        expect(() => {
          runChecks(builder, {});
        }).not.toThrow();
      });

      it('errors on non-objects', () => {
        builder.type = 'object';

        expect(() => {
          runChecks(builder, 123);
        }).toThrowErrorMatchingSnapshot();
      });

      it('errors on arrays', () => {
        builder.type = 'object';

        expect(() => {
          runChecks(builder, []);
        }).toThrowErrorMatchingSnapshot();
      });

      it('errors on nulls', () => {
        builder.type = 'object';

        expect(() => {
          runChecks(builder, null);
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('string', () => {
      it('allows strings', () => {
        builder.type = 'string';

        expect(() => {
          runChecks(builder, 'foo');
        }).not.toThrow();
      });

      it('errors on non-strings', () => {
        builder.type = 'string';

        expect(() => {
          runChecks(builder, 123);
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('custom()', () => {
    it('errors if no callback', () => {
      // @ts-ignore
      expect(() => builder.custom()).toThrow('Custom blueprints require a validation function.');
    });

    it('errors if callback is not a function', () => {
      // @ts-ignore
      expect(() => builder.custom(123)).toThrow('Custom blueprints require a validation function.');
    });

    it('triggers callback function', () => {
      builder = custom(value => {
        if (value === 123) {
          throw new Error('This will error!');
        }
      }, 0);

      expect(() => {
        runChecks(builder, 123);
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        runChecks(builder, 456);
      }).not.toThrow();
    });

    it('is passed entire options object', () => {
      builder = custom<unknown, { foo?: number; bar?: number }>((value, options) => {
        if (options.foo && options.bar) {
          throw new Error('This will error!');
        }
      }, 0);

      expect(() => {
        builder.runChecks('error', 123, { foo: 123, bar: 456, error: '' });
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('invariant()', () => {
    it('does nothing if condition is true', () => {
      expect(() => {
        builder.invariant(true, '');
      }).not.toThrow();
    });

    it('errors if condition is false', () => {
      expect(() => {
        builder.invariant(false, 'Failure');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes an option path', () => {
      expect(() => {
        builder.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a class name', () => {
      expect(() => {
        // @ts-ignore Allow access
        builder.options.name = 'FooBar';

        builder.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a class name when no path', () => {
      expect(() => {
        // @ts-ignore Allow access
        builder.options.name = 'FooBar';

        builder.invariant(false, 'Failure');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a file name', () => {
      expect(() => {
        // @ts-ignore Allow access
        builder.options.file = 'package.json';

        builder.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a file and class name', () => {
      expect(() => {
        // @ts-ignore Allow access
        builder.options.file = 'package.json';
        // @ts-ignore Allow access
        builder.options.name = 'FooBar';

        builder.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('key()', () => {
    it('returns as-is if not deep', () => {
      // @ts-ignore Allow access
      expect(builder.key('foo')).toBe('foo');
    });

    it('returns last part of a deep path', () => {
      // @ts-ignore Allow access
      expect(builder.key('foo.bar.baz')).toBe('baz');
    });
  });

  describe('message()', () => {
    it('errors for empty value', () => {
      expect(() => {
        builder.message('');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors for non-string value', () => {
      expect(() => {
        // @ts-ignore
        builder.message(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('sets message', () => {
      builder.message('foobar');

      // @ts-ignore Allow access
      expect(builder.errorMessage).toBe('foobar');
    });
  });

  describe('deprecate()', () => {
    it('errors for empty value', () => {
      expect(() => {
        builder.deprecate('');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors for non-string value', () => {
      expect(() => {
        // @ts-ignore
        builder.deprecate(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('sets message', () => {
      builder.deprecate('foobar');

      // @ts-ignore Allow access
      expect(builder.deprecatedMessage).toBe('foobar');
    });
  });

  describe('nullable()', () => {
    it('errors if field is undefined', () => {
      expect(() => {
        builder.notNullable();

        runChecks(builder, null);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if field is defined', () => {
      expect(() => {
        builder.nullable();

        runChecks(builder, null);
      }).not.toThrow();
    });
  });

  describe('required()', () => {
    it('errors if field is undefined', () => {
      expect(() => {
        builder.required();

        runChecks(builder);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if field is defined', () => {
      expect(() => {
        builder.required(false);

        runChecks(builder, 'foo');
      }).not.toThrow();
    });
  });

  describe('runChecks()', () => {
    it('returns valid value', () => {
      expect(builder.runChecks('key', 'bar', {})).toBe('bar');
    });

    it('returns default value if value passed is undefined', () => {
      expect(builder.runChecks('key', undefined, {})).toBe('foo');
    });

    it('returns null if value passed is null and builder is nullable', () => {
      expect(builder.nullable().runChecks('key', null, {})).toBeNull();
    });

    it('errors if value passed is undefined and builder is required', () => {
      expect(() => {
        builder.required().runChecks('key', undefined, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value passed is null and builder is non-nullable', () => {
      expect(() => {
        builder.runChecks('key', null, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value passed is defined and builder is never', () => {
      expect(() => {
        // @ts-ignore Allow values
        expect(builder.never().runChecks('key', 'bar', { key: 'bar' })).toBeUndefined();
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs default type of check', () => {
      expect(() => {
        builder.runChecks('key', 123, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('uses custom message', () => {
      expect(() => {
        builder.message('Oops, something is broken.').runChecks('key', 123, {});
      }).toThrowErrorMatchingSnapshot();
    });

    describe('deprecation', () => {
      const oldInfo = console.info;

      beforeEach(() => {
        jest.spyOn(console, 'info').mockImplementation();
      });

      afterEach(() => {
        console.info = oldInfo;
      });

      it('logs a message', () => {
        builder.deprecate('Use something else.');
        builder.runChecks('key', 'foo', {});

        expect(console.info).toHaveBeenCalledWith('Field "key" is deprecated. Use something else.');
      });

      it('doesnt log if undefined', () => {
        builder.deprecate('Use something else.');
        builder.runChecks('key', undefined, {});

        expect(console.info).not.toHaveBeenCalledWith(
          'Field "key" is deprecated. Use something else.',
        );
      });
    });
  });

  describe('only()', () => {
    beforeEach(() => {
      builder.only();
    });

    it('errors if default value is not the same type', () => {
      builder.defaultValue = 123;

      expect(() => {
        builder.only();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value doesnt match the default value', () => {
      expect(() => {
        runChecks(builder, 'bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value matches default value', () => {
      expect(() => {
        runChecks(builder, 'foo');
      }).not.toThrow();
    });
  });

  describe('or()', () => {
    beforeEach(() => {
      builder.or('bar', 'baz');
    });

    it('errors if no keys are defined', () => {
      expect(() => {
        builder.or();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if not 1 option is defined', () => {
      expect(() => {
        runChecks(builder, 'a', { key: 'foo', struct: {} });
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if at least 1 option is defined', () => {
      expect(() => {
        runChecks(builder, 'a', { key: 'foo', struct: { foo: 'a' } });
      }).not.toThrow();
    });

    it('doesnt error if at least 1 option is defined that isnt the main field', () => {
      expect(() => {
        runChecks(builder, 'a', { key: 'foo', struct: { bar: 'b' } });
      }).not.toThrow();
    });
  });

  describe('xor()', () => {
    beforeEach(() => {
      builder.xor('bar', 'baz');
    });

    it('errors if no keys are defined', () => {
      expect(() => {
        builder.xor();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if no options are defined', () => {
      expect(() => {
        runChecks(builder, 'a', { key: 'foo', struct: {} });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if more than 1 option is defined', () => {
      expect(() => {
        runChecks(builder, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
            bar: 'b',
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if only 1 option is defined', () => {
      expect(() => {
        runChecks(builder, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
          },
        });
      }).not.toThrow();
    });
  });
});

describe('custom()', () => {
  it('returns a builder', () => {
    expect(custom(() => {}, '')).toBeInstanceOf(Builder);
  });

  it('sets type and default value', () => {
    const builder = custom(() => {}, 123);

    expect(builder.type).toBe('custom');
    expect(builder.defaultValue).toBe(123);
  });

  it('returns the type alias', () => {
    expect(custom(() => {}, '').typeAlias()).toBe('custom');
  });
});

describe('func()', () => {
  it('returns a builder', () => {
    expect(func()).toBeInstanceOf(Builder);
  });

  it('sets type and default value', () => {
    const noop = () => {};
    const builder = func(noop);

    expect(builder.type).toBe('function');
    expect(builder.defaultValue).toBe(noop);
  });

  it('errors if a non-function value is used', () => {
    expect(() => {
      func().runChecks(
        'key',
        // @ts-ignore Test invalid type
        123,
        {},
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('returns the type alias', () => {
    expect(func().typeAlias()).toBe('function');
  });
});
