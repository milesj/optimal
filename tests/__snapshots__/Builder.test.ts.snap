// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder and() errors if no keys are defined 1`] = `"AND requires a list of field names."`;

exports[`Builder and() errors if not all options are defined 1`] = `"All of these fields must be defined: foo, bar, baz"`;

exports[`Builder checkType() array errors on non-arrays 1`] = `"Invalid field \\"key\\". Must be an array."`;

exports[`Builder checkType() boolean errors on non-booleans 1`] = `"Invalid field \\"key\\". Must be a boolean."`;

exports[`Builder checkType() function errors on non-functions 1`] = `"Invalid field \\"key\\". Must be a function."`;

exports[`Builder checkType() number errors on non-numbers 1`] = `"Invalid field \\"key\\". Must be a number."`;

exports[`Builder checkType() object errors on arrays 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`Builder checkType() object errors on non-objects 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`Builder checkType() object errors on nulls 1`] = `"Invalid field \\"key\\". Null is not allowed."`;

exports[`Builder checkType() string errors on non-strings 1`] = `"Invalid field \\"key\\". Must be a string."`;

exports[`Builder custom() is passed entire options object 1`] = `"Invalid field \\"error\\". This will error!"`;

exports[`Builder custom() triggers callback function 1`] = `"Invalid field \\"key\\". This will error!"`;

exports[`Builder deprecate() errors for empty value 1`] = `"A non-empty string is required for deprecated messages."`;

exports[`Builder deprecate() errors for non-string value 1`] = `"A non-empty string is required for deprecated messages."`;

exports[`Builder errors if default value is undefined 1`] = `"A default value for type \\"string\\" is required."`;

exports[`Builder invariant() errors if condition is false 1`] = `"Failure"`;

exports[`Builder invariant() includes a class name 1`] = `"Invalid FooBar field \\"foo.bar\\". Failure"`;

exports[`Builder invariant() includes a class name when no path 1`] = `"FooBar. Failure"`;

exports[`Builder invariant() includes a file and class name 1`] = `"Invalid FooBar field \\"foo.bar\\" in package.json. Failure"`;

exports[`Builder invariant() includes a file name 1`] = `"Invalid field \\"foo.bar\\" in package.json. Failure"`;

exports[`Builder invariant() includes an option path 1`] = `"Invalid field \\"foo.bar\\". Failure"`;

exports[`Builder message() errors for empty value 1`] = `"A non-empty string is required for custom messages."`;

exports[`Builder message() errors for non-string value 1`] = `"A non-empty string is required for custom messages."`;

exports[`Builder nullable() errors if field is undefined 1`] = `"Invalid field \\"key\\". Null is not allowed."`;

exports[`Builder only() errors if default value is not the same type 1`] = `"Only requires a default string value."`;

exports[`Builder only() errors if value doesnt match the default value 1`] = `"Invalid field \\"key\\". Value may only be \\"foo\\"."`;

exports[`Builder or() errors if no keys are defined 1`] = `"OR requires a list of field names."`;

exports[`Builder or() errors if not 1 option is defined 1`] = `"At least one of these fields must be defined: foo, bar, baz"`;

exports[`Builder required() errors if field is undefined 1`] = `"Invalid field \\"key\\". Field is required and must be defined."`;

exports[`Builder runChecks() errors if value passed is defined and builder is never 1`] = `"Invalid field \\"key\\". Field should never be used."`;

exports[`Builder runChecks() errors if value passed is null and builder is non-nullable 1`] = `"Invalid field \\"key\\". Null is not allowed."`;

exports[`Builder runChecks() errors if value passed is undefined and builder is required 1`] = `"Invalid field \\"key\\". Field is required and must be defined."`;

exports[`Builder runChecks() runs default type of check 1`] = `"Invalid field \\"key\\". Must be a string."`;

exports[`Builder runChecks() uses custom message 1`] = `"Invalid field \\"key\\". Oops, something is broken."`;

exports[`Builder xor() errors if more than 1 option is defined 1`] = `"Only one of these fields may be defined: foo, bar, baz"`;

exports[`Builder xor() errors if no keys are defined 1`] = `"XOR requires a list of field names."`;

exports[`Builder xor() errors if no options are defined 1`] = `"Only one of these fields may be defined: foo, bar, baz"`;

exports[`func() errors if a non-function value is used 1`] = `"Invalid field \\"key\\". Must be a function."`;
